### PR TITLE
Remove IndexedInSubqueryFilter, add SemiLookupJoin, AntiJoin

### PR DESCRIPTION
**Pull request by max-hoffman**
- IN_SUBQUERY and EQUALS_SUBQUERY (and NOT ...) are converted to joins when the subquery does not reference the outer scope
- RIGHT_SEMI_JOIN and RIGHT_SEMI_LOOKUP_JOIN are added as a replacement for IndexedInSubqueryFilter
- MAX_1_ROW added for runtime-enforcement, previously handled by EvalMultiple
- fix various bugs in join iter and lookup generation

Note:
- RIGHT_SEMI_LOOKUP_JOIN is only safe when the left side of the join is converted into a unique index lookup, otherwise duplicate rows will have been missed

TODO:
- selection, projection, etc rearrangements moving towards JoinApply for correlated expressions
- collapse scopes without losing MAX_1_ROW enforcement
- deduplicate tables in joins without losing MAX_1_ROW and row duplication constraints

### CODE REVIEW

File: enginetest/queries/query_plans.go [hunk 1]
```diff
@@ -3295,29 +3415,45 @@ inner join pq on true
...
-			" │   ├─ cacheable: true\n" +
-			" │   └─ Limit(1)\n" +
-			" │       └─ Table\n" +
-			" │           ├─ name: othertable\n" +
-			" │           └─ columns: [i2]\n" +
-			" └─ IndexedTableAccess(mytable)\n" +
-			"     └─ index: [mytable.i]\n" +
+		ExpectedPlan: "Project\n" +
+			" ├─ columns: [mytable.i:1!null, mytable.s:2!null]\n" +
+			" └─ LookupJoin\n" +
```
Comment (max-hoffman) [inline enginetest/queries/query_plans.go:3420]: todo this should be semi

File: enginetest/queries/query_plans.go [hunk 2]
```diff
@@ -11400,17 +11610,19 @@ var IntegrationPlanTests = []QueryPlanTest{
...
-			" │               │   ├─ left: E2I7U.id:17!null\n" +
-			" │               │   └─ right: Subquery\n" +
-			" │               │       ├─ cacheable: true\n" +
-			" │               │       └─ Table\n" +
-			" │               │           ├─ name: AMYXQ\n" +
-			" │               │           └─ columns: [luevy]\n" +
-			" │               │  ))\n" +
-			" │               └─ Table\n" +
-			" │                   └─ name: E2I7U\n" +
+			" │           └─ AntiJoin\n" +
```
Comment (max-hoffman) [inline enginetest/queries/query_plans.go:11613]: todo: maybe we need antiHashJoin? which is kind of going to be the original plan, but at least we can choose between options

File: enginetest/queries/query_plans.go [hunk 3]
```diff
@@ -12602,523 +12828,531 @@ WHERE
...
+			"     ├─ Eq\n" +
+			"     │   ├─ TUPLE(fs.T4IBQ:0!null, fs.M6T2N:1, fs.BTXC5:2, fs.TUV25:3)\n" +
+			"     │   └─ TUPLE(applySubq0.T4IBQ:4!null, applySubq0.M6T2N:5, applySubq0.BTXC5:6, applySubq0.TUV25:7)\n" +
+			"     ├─ SubqueryAlias\n" +
+			"     │   ├─ name: fs\n" +
+			"     │   ├─ outerVisibility: false\n" +
+			"     │   ├─ cacheable: true\n" +
+			"     │   └─ Project\n" +
+			"     │       ├─ columns: [RSA3Y.T4IBQ:3!null as T4IBQ, JMHIE.M6T2N:0 as M6T2N, JMHIE.BTXC5:1 as BTXC5, JMHIE.TUV25:2 as TUV25]\n" +
+			"     │       └─ CrossJoin\n" +
```
Comment (max-hoffman) [inline enginetest/queries/query_plans.go:12841]: hmm, this seems bad

File: sql/plan/join_iters.go [hunk 4]
```diff
@@ -268,11 +268,14 @@ func (i *existsIter) Next(ctx *sql.Context) (sql.Row, error) {
...
 				}
 				if errors.Is(err, io.EOF) {
-					if i.typ == JoinTypeSemi {
+					if i.typ.IsSemi() {
 						// reset iter, no match
 						break
 					}
-					return left, nil
+					if i.typ.IsRightPartial() {
+						return right, nil
```
Comment (max-hoffman) [inline sql/plan/join_iters.go:276]: this probably needs scope prepended

File: enginetest/queries/query_plans.go [hunk 5]
```diff
@@ -8495,84 +8689,76 @@ var IntegrationPlanTests = []QueryPlanTest{
 	       PV6R5.NUMK2 <> 1
 	`,
 		ExpectedPlan: "Project\n" +
-			" ├─ columns: [rn.id:44!null as id, concat(NSPLT.TW55N:13!null,FDNCN (longtext),LQNCX.TW55N:30!null) as X37NA, concat(XLZA5.TW55N:63!null,FDNCN (longtext),AFJMD.TW55N:80!null) as THWCS, rn.HVHRZ:47!null as HVHRZ]\n" +
+			" ├─ columns: [rn.id:44!null as id, concat(NSPLT.TW55N:3!null,FDNCN (longtext),LQNCX.TW55N:30!null) as X37NA, concat(XLZA5.TW55N:53!null,FDNCN (longtext),AFJMD.TW55N:80!null) as THWCS, rn.HVHRZ:47!null as HVHRZ]\n" +
```
Comment (max-hoffman) [inline enginetest/queries/query_plans.go:8692]: changes unrelated to new operators are a result of casing bugs blocking lookupJoins

**Review by reltuk — APPROVED** [review]
Generally looks quite good. As discussed offline, excluding the dual table case makes me suspicious – seems like that should work with all this machinery without issue.

**Pull request merged**